---
title: "Case/Control GWAS for Age-related Macular Degeneration"
author: "Andrea Badger, Hanley Kingston, Seth Temple"
date: "December 3, 2019"
output:
  pdf_document: default
  html_document: default
---

First, we unzip the AMD.txt.gz compressed file with software available online. Then we read this data into a R data frame.
```{r}
amd.df <- read.table('~/UW/Student/STAT 551/H/H3/AMD.txt', header = T)
amd.df <- subset(amd.df, chromosome != 'X') # remove sex chromosomes
amd.df <- subset(amd.df, chromosome != 'X_random')
amd.df <- subset(amd.df, chromosome != 'Un_random') # remove these < 20 SNPs
```

Convert data to refer to dosages.
```{r}
amd.geno <- t(data.matrix(amd.df[,5:150])) - 1
```

Create a matrix of genotype counts.
```{r}
amd.hwe <- data.frame(matrix(0, nrow = ncol(amd.geno), ncol = 4))
colnames(amd.hwe) <- c(-1,0,1,2)
for(i in 1:ncol(amd.geno)){
  cts <- summary(as.factor(amd.geno[,i]))
  nms <- names(cts)
  for(j in 1:length(nms)){
    amd.hwe[i,nms[j]] <- unname(cts[j])
  }
}
```

Remove SNPs that have a less than 85% call rate.
```{r}
ct <- nrow(amd.hwe)
bound <- .15 * nrow(amd.geno)
amd.hwe$callrt <- amd.hwe$`-1` < bound
amd.hwe <- subset(amd.hwe, callrt == T)
nrow(amd.hwe) - ct
```

Remove monomorphic SNPs.
```{r}
ct <- nrow(amd.hwe)
amd.hwe$mono <- amd.hwe$`0` == nrow(amd.geno) | amd.hwe$`2` == nrow(amd.geno)
amd.hwe <- subset(amd.hwe, mono == F)
nrow(amd.hwe) - ct
```

Remove SNPs lacking heterozogosity.
```{r}
ct <- nrow(amd.hwe)
amd.hwe$hom <- (amd.hwe$`0` + amd.hwe$`2`) == nrow(amd.geno)
amd.hwe <- subset(amd.hwe, hom == F)
nrow(amd.hwe) - ct
```

Remove SNPs according to a MAF rule.
```{r}
ct <- nrow(amd.hwe)
mbound <- .01 * nrow(amd.geno) * 2
amd.hwe$maf <- (2 * amd.hwe$`0` + amd.hwe$`1`) < mbound | (2 * amd.hwe$`2` + amd.hwe$`1`) < mbound
amd.hwe <- subset(amd.hwe, maf == F)
nrow(amd.hwe) - ct
```

We will use Graffelmann's package to perform HWE testing.
```{r}
library(HardyWeinberg)
amd.hwe <- data.matrix(amd.hwe[,2:4])
colnames(amd.hwe) <- c('AA','AB','BB')
```

Only keep SNPs that pass Hardy Weinberg quality control.
```{r}
ct <- nrow(amd.hwe)
p_bonf <- .05 / nrow(amd.hwe) # bonferroni correction for multiple testing
p <- suppressWarnings(HWChisqMat(amd.hwe, verbose = F))$pvalvec
keep <- p > p_bonf
amd.hwe <- amd.hwe[keep == T,]
nrow(amd.hwe) - ct
```

\newpage
```{r}
plot(-log10(seq(0,1,length.out = length(p))), sort(-log10(p), decreasing = T),
     pch = 20,
     main = 'QQ Plot for Hardy Weinberg QC',
     ylab = 'p-value   (-log10)',
     xlab = 'p-value   (-log10)')
abline(0, 1, col = 'red')
```
\newpage
```{r}
HWTernaryPlot(amd.hwe, alpha = p_bonf)
```

```{r}
# apply data cleaning to dosage matrix
cnames <- colnames(amd.geno)
rnames <- rownames(amd.hwe)
keep <- cnames %in% rnames

amd.geno <- amd.geno[, keep == T]
ncol(amd.geno)
```
Note that the data cleaning closely aligns with that of the Klein (2005) paper and supplement.

\newpage
Prepare data to perform dichotomous trait case/control statistical testing.
```{r}
amd.case <- data.frame(matrix(0, nrow = ncol(amd.geno), ncol = 4))
amd.control <- data.frame(matrix(0, nrow = ncol(amd.geno), ncol = 4))
colnames(amd.case) <- c(-1,0,1,2)
colnames(amd.control) <- c(-1,0,1,2)
# create genotype count data frame for cases
for(i in 1:ncol(amd.geno)){
  cts <- summary(as.factor(amd.geno[1:96,i]))
  nms <- names(cts)
  for(j in 1:length(nms)){
    amd.case[i,nms[j]] <- unname(cts[j])
  }
}
# create genotype count data frame for controls
for(i in 1:ncol(amd.geno)){
  cts <- summary(as.factor(amd.geno[97:146,i]))
  nms <- names(cts)
  for(j in 1:length(nms)){
    amd.control[i,nms[j]] <- unname(cts[j])
  }
}
# convert data frames to matrices
amd.case <- data.matrix(amd.case[,2:4])
amd.control <- data.matrix(amd.control[,2:4])
```

Calculate contingency tables and perform chi-squared tests with Bonferroni correction.
```{r}
# see lecture notes
num <- nrow(amd.case)
X <- seq(0, length.out = num)
for(i in 1:num){
  
  # defining variables in the contingency table
  R <- sum(amd.case[i,])
  S <- sum(amd.control[i,])
  N <- R + S
  r0 <- amd.case[i,1]
  r1 <- amd.case[i,2]
  r2 <- amd.case[i,3]
  s0 <- amd.control[i,1]
  s1 <- amd.control[i,2]
  s2 <- amd.control[i,3]
  n0 <- r0 + s0
  n1 <- r1 + s1
  n2 <- r2 + s2
  X_num <- 2 * N * (( N * (r1 + 2 * r2) - R * (n1 + 2 * n2)) ^ 2)
  X_den <- S * R * (2 * N * (n1 + 2 * n2) - (n1 + 2 * n2) ^ 2)
  X[i] <- X_num / X_den
  
}
# convert chi-squared values to p-values
P <- pchisq(X, 1, lower.tail = F)
```

\newpage
```{r}
plot(-log10(P),
     main = 'Manhattan Plot for Age-related Macular Degeneration',
     ylab = 'p-value   (-log10)',
     xlab = 'Indices',
     pch = 20)
abline(-log10(5e-8), 0, col = 'green') # standard significance level for GWAS
abline(-log10(.05 / nrow(amd.case)), 0, col = 'red') # .05 significance with Bonferroni correction
abline(-log10(.10 / nrow(amd.case)), 0, col = 'blue') # .10 significance with Bonferroni correction
```

\newpage
We find the SNP with the lowest p-value. (There is a typo in the Klein (2005) paper.)
```{r}
Pfind <- P
id <- which(Pfind == min(Pfind), arr.ind = TRUE)
Pfind[id]
Pfind[id] <- 1
ID <- as.integer(rownames(amd.hwe)[id])
amd.df$rsID[ID]
```
Below are the genotype counts for statistically significant SNP we found.
```{r}
unname(amd.case[id,])
unname(amd.control[id,])
sum(amd.hwe[id,]) # successful calls
```

We find the SNP with the 2nd lowest p-value.
```{r}
id <- which(Pfind == min(Pfind), arr.ind = TRUE)
Pfind[id]
Pfind[id] <- 1
ID <- as.integer(rownames(amd.hwe)[id])
amd.df$rsID[ID]
```
The data we have has rs1329428 genotyped again as mentioned in Klein (2005).
```{r}
unname(amd.case[id,])
unname(amd.control[id,])
sum(amd.hwe[id,]) # successful calls
```

